pkg/api/password.go

```diff
@@ -27,27 +27,33 @@ func SendResetPasswordEmail(c *models.ReqContext) response.Response {
 
 	userQuery := models.GetUserByLoginQuery{LoginOrEmail: form.UserOrEmail}
 
-	if err := bus.Dispatch(c.Req.Context(), &userQuery); err != nil {
+	if err := hs.SQLStore.GetUserByLogin(c.Req.Context(), &userQuery); err != nil {
 		c.Logger.Info("Requested password reset for user that was not found", "user", userQuery.LoginOrEmail)
 		return response.Error(200, "Email sent", err)
 	}
 
 	emailCmd := models.SendResetPasswordEmailCommand{User: userQuery.Result}
-	if err := bus.Dispatch(c.Req.Context(), &emailCmd); err != nil {
+	if err := hs.NotificationService.SendResetPasswordEmail(c.Req.Context(), &emailCmd); err != nil {
 		return response.Error(500, "Failed to send email", err)
 	}
 
 	return response.Success("Email sent")
 }
 
-func ResetPassword(c *models.ReqContext) response.Response {
+func (hs *HTTPServer) ResetPassword(c *models.ReqContext) response.Response {
 	form := dtos.ResetUserPasswordForm{}
 	if err := web.Bind(c.Req, &form); err != nil {
 		return response.Error(http.StatusBadRequest, "bad request data", err)
 	}
 	query := models.ValidateResetPasswordCodeQuery{Code: form.Code}
 
-	if err := bus.Dispatch(c.Req.Context(), &query); err != nil {
+	getUserByLogin := func(ctx context.Context, login string) (*models.User, error) {
+		userQuery := models.GetUserByLoginQuery{LoginOrEmail: login}
+		err := hs.SQLStore.GetUserByLogin(ctx, &userQuery)
+		return userQuery.Result, err
+	}
+
+	if err := hs.NotificationService.ValidateResetPasswordCode(c.Req.Context(), &query, getUserByLogin); err != nil {
```

Comment:
This is a place where I think a separate controller would be helpful. It's not NotificationService who should do the code validation, and it's not the API handler who should look up users while resetting the password. It should be a separate controller, called from the API and using Store+NotificationService to do the job.

But I leave it to further generations to improve.